src/WebJobs.Script/Host/FunctionMetadataManager.cs

```diff
@@ -18,36 +22,91 @@ namespace Microsoft.Azure.WebJobs.Script
+        private ILogger _logger;
+        private IOptions<ScriptJobHostOptions> _scriptOptions;
+        private ImmutableArray<FunctionMetadata> _functionMetadataFilteredArray;
+        private ImmutableArray<FunctionMetadata> _functionMetadataAllArray;
+        private IEnumerable<IFunctionProvider> _functionProviders;
         private Dictionary<string, ICollection<string>> _functionErrors = new Dictionary<string, ICollection<string>>();
 
-        public FunctionMetadataManager(IOptions<ScriptJobHostOptions> scriptOptions, IFunctionMetadataProvider functionMetadataProvider, IOptions<HttpWorkerOptions> httpWorkerOptions, ILoggerFactory loggerFactory)
+        public FunctionMetadataManager(IOptions<ScriptJobHostOptions> scriptOptions, IFunctionMetadataProvider functionMetadataProvider, IEnumerable<IFunctionProvider> functionProviders, IOptions<HttpWorkerOptions> httpWorkerOptions, IScriptHostManager scriptHostManager, ILoggerFactory loggerFactory)
         {
             _scriptOptions = scriptOptions;
+            _serviceProvider = scriptHostManager as IServiceProvider;
             _functionMetadataProvider = functionMetadataProvider;
             _logger = loggerFactory.CreateLogger(LogCategories.Startup);
-            _functionMetadataArray = new Lazy<ImmutableArray<FunctionMetadata>>(LoadFunctionMetadata);
-            _isHttpWorker = httpWorkerOptions.Value.Description != null;
-        }
+            _isHttpWorker = httpWorkerOptions?.Value?.Description != null;
+            _functionProviders = functionProviders;
 
-        public ImmutableArray<FunctionMetadata> Functions => _functionMetadataArray.Value;
+            scriptHostManager.HostInitializing += ResetScriptHostServices;
+        }
 
         public ImmutableDictionary<string, ImmutableArray<string>> Errors { get; private set; }
 
+        /// <summary>
+        /// Gets the function metadata array from all providers.
+        /// </summary>
+        /// <param name="forceRefresh">Forces reload from all providers.</param>
+        /// <param name="includeBlocked">Includes all functions, including blocked ones.</param>
+        /// <returns>An Immmutable array of FunctionMetadata.</returns>
+        public ImmutableArray<FunctionMetadata> GetFunctionMetadata(bool forceRefresh, bool includeBlocked = false)
+        {
+            var currentlyLoaded = GetLoadedMetadata(includeBlocked);
+            if (!forceRefresh && !_servicesReset && !currentlyLoaded.IsDefaultOrEmpty)
+            {
+                return currentlyLoaded;
+            }
+
+            LoadFunctionMetadata(forceRefresh);
+            _servicesReset = false;
+
+            return GetLoadedMetadata(includeBlocked);
+        }
+
+        internal ImmutableArray<FunctionMetadata> GetLoadedMetadata(bool includeBlocked)
+        {
+            return includeBlocked ? _functionMetadataAllArray : _functionMetadataFilteredArray;
+        }
+
+        private void ResetScriptHostServices(object sender, EventArgs e)
```

Comment:
Rename this to just Reset, and call this from the ctor to do the init. That way we have a single place we're maintaining this.

Reply:
I had it `Reset` initially, but Fabio made a good point that we should be clearer and explicitly call out in the name that we are resetting ScripHost level services, and suggested the name `ResetScriptHostServices`. I think I like that it's clear specially if it's only a private function and can call out specifics.

Reply:
Also calling initialize in ctor is a bit tricky because the Script Host level service provider doesn't yet have an logging initialized and I get a null exception. So, until Script Host is initializing, we can't start getting services from it's service provider.

Maybe I should add comments about that, the fact that these services are changing scopes,